Unable to compile Kokkos_SIMD for BDW arch using GCC 8.5.0

I get the following error when trying to compile Kokkos_SIMD using GCC 8.5.0 on a Broadwell Architecture. The error might be reproducible on all AVX2 archs. 

- Compilers : GCC 8.5.0
- Kokkos version: Kokkos 4.0.0 (https://github.com/kokkos/kokkos/archive/refs/tags/4.0.00.zip)
- Platform and backend: Tried compiling for Broadwell architecture.
- cmake command: cmake3 -S . -B build -DCMAKE_BUILD_TYPE=Release -DKokkos_ARCH_BDW=ON
- cmake build output: 
```
In file included from /home/rk9481/cicc_issue_kokkos/build/_deps/kokkos-src/simd/src/Kokkos_SIMD.hpp:25,
                 from /home/rk9481/cicc_issue_kokkos/main.cpp:2:
/home/rk9481/cicc_issue_kokkos/build/_deps/kokkos-src/simd/src/Kokkos_SIMD_AVX2.hpp: In member function ‘void Kokkos::Experimental::simd<long int, Kokkos::Experimental::simd_abi::avx2_fixed_size<4> >::copy_from(const value_type*, Kokkos::Experimental::element_aligned_tag)’:
/home/rk9481/cicc_issue_kokkos/build/_deps/kokkos-src/simd/src/Kokkos_SIMD_AVX2.hpp:730:37: error: invalid conversion from ‘const value_type*’ {aka ‘const long int*’} to ‘const long long int*’ [-fpermissive]
     m_value = _mm256_maskload_epi64(ptr, static_cast<__m256i>(mask_type(true)));
                                     ^~~
In file included from /usr/lib/gcc/x86_64-redhat-linux/8/include/immintrin.h:43,
                 from /home/rk9481/cicc_issue_kokkos/build/_deps/kokkos-src/simd/src/Kokkos_SIMD_AVX2.hpp:25,
                 from /home/rk9481/cicc_issue_kokkos/build/_deps/kokkos-src/simd/src/Kokkos_SIMD.hpp:25,
                 from /home/rk9481/cicc_issue_kokkos/main.cpp:2:
/usr/lib/gcc/x86_64-redhat-linux/8/include/avx2intrin.h:1126:41: note:   initializing argument 1 of ‘__m256i _mm256_maskload_epi64(const long long int*, __m256i)’
 _mm256_maskload_epi64 (long long const *__X, __m256i __M )
                        ~~~~~~~~~~~~~~~~~^~~
In file included from /home/rk9481/cicc_issue_kokkos/build/_deps/kokkos-src/simd/src/Kokkos_SIMD.hpp:25,
                 from /home/rk9481/cicc_issue_kokkos/main.cpp:2:
/home/rk9481/cicc_issue_kokkos/build/_deps/kokkos-src/simd/src/Kokkos_SIMD_AVX2.hpp: In member function ‘void Kokkos::Experimental::simd<long int, Kokkos::Experimental::simd_abi::avx2_fixed_size<4> >::copy_to(Kokkos::Experimental::simd<long int, Kokkos::Experimental::simd_abi::avx2_fixed_size<4> >::value_type*, Kokkos::Experimental::element_aligned_tag) const’:
/home/rk9481/cicc_issue_kokkos/build/_deps/kokkos-src/simd/src/Kokkos_SIMD_AVX2.hpp:734:28: error: invalid conversion from ‘Kokkos::Experimental::simd<long int, Kokkos::Experimental::simd_abi::avx2_fixed_size<4> >::value_type*’ {aka ‘long int*’} to ‘long long int*’ [-fpermissive]
     _mm256_maskstore_epi64(ptr, static_cast<__m256i>(mask_type(true)), m_value);
                            ^~~
In file included from /usr/lib/gcc/x86_64-redhat-linux/8/include/immintrin.h:43,
                 from /home/rk9481/cicc_issue_kokkos/build/_deps/kokkos-src/simd/src/Kokkos_SIMD_AVX2.hpp:25,
                 from /home/rk9481/cicc_issue_kokkos/build/_deps/kokkos-src/simd/src/Kokkos_SIMD.hpp:25,
                 from /home/rk9481/cicc_issue_kokkos/main.cpp:2:
/usr/lib/gcc/x86_64-redhat-linux/8/include/avx2intrin.h:1157:36: note:   initializing argument 1 of ‘void _mm256_maskstore_epi64(long long int*, __m256i, __m256i)’
 _mm256_maskstore_epi64 (long long *__X, __m256i __M, __m256i __Y )
                         ~~~~~~~~~~~^~~
gmake[2]: *** [CMakeFiles/cicc_test.dir/build.make:76: CMakeFiles/cicc_test.dir/main.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:280: CMakeFiles/cicc_test.dir/all] Error 2
gmake: *** [Makefile:136: all] Error 2
```
- Code:
```
  #include <Kokkos_Core.hpp>
#include <Kokkos_SIMD.hpp>
#include "small_mat_mul.hpp"

int main(){

    Kokkos::initialize();
    {
    }
    Kokkos::finalize();

    return 1;
}
```
I was using Kokkos_SIMD for another project and ran into this compilation error, so I tried to reproduce the error using a bare bones Kokkos code.

--- a/simd/src/Kokkos_SIMD_AVX2.hpp
+++ b/simd/src/Kokkos_SIMD_AVX2.hpp
@@ -680,10 +680,12 @@ template <>
 class simd<std::int64_t, simd_abi::avx2_fixed_size<4>> {
   __m256i m_value;
 
+  static_assert(sizeof(long long) == 8);
+
  public:
   using value_type = std::int64_t;
   using abi_type   = simd_abi::avx2_fixed_size<4>;
-  using mask_type  = simd_mask<value_type, abi_type>;
+  using mask_type  = simd_mask<std::int64_t, abi_type>;
   using reference  = value_type&;
   KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION simd()            = default;
   KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION simd(simd const&) = default;
@@ -727,11 +729,13 @@ class simd<std::int64_t, simd_abi::avx2_fixed_size<4>> {
   }
   KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION void copy_from(value_type const* ptr,
                                                        element_aligned_tag) {
-    m_value = _mm256_maskload_epi64(ptr, static_cast<__m256i>(mask_type(true)));
+    m_value = _mm256_maskload_epi64(reinterpret_cast<long long const*>(ptr),
+                                    static_cast<__m256i>(mask_type(true)));
   }
   KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION void copy_to(
       value_type* ptr, element_aligned_tag) const {
-    _mm256_maskstore_epi64(ptr, static_cast<__m256i>(mask_type(true)), m_value);
+    _mm256_maskstore_epi64(reinterpret_cast<long long*>(ptr),
+                           static_cast<__m256i>(mask_type(true)), m_value);
   }
   KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION constexpr explicit operator __m256i()
       const {

--- a/simd/src/Kokkos_SIMD_AVX2.hpp
+++ b/simd/src/Kokkos_SIMD_AVX2.hpp
@@ -685,7 +685,7 @@ class simd<std::int64_t, simd_abi::avx2_fixed_size<4>> {
  public:
   using value_type = std::int64_t;
   using abi_type   = simd_abi::avx2_fixed_size<4>;
-  using mask_type  = simd_mask<std::int64_t, abi_type>;
+  using mask_type  = simd_mask<value_type, abi_type>;
   using reference  = value_type&;
   KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION simd()            = default;
   KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION simd(simd const&) = default;

